### PR TITLE
Ext_Authz duration dynamic metadata

### PR DESCRIPTION
**Pull request by llu94**
Risk Level: low
Testing: unit testing
Docs Changes: n/a
Release Notes: inline

Issue Link: #19255

### CODE REVIEW

File: api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto
Comment (llu94) [reply api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:394]: I was told it's probably because my underlying Clang version is not in line with what Envoy uses so I will be re-installing Clang later. Regardless I actually can't even push past the format pre-checks from local to remote without doing this.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (llu94) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:5]: They are not, good catch. Left over from a previous attempt.

File: source/extensions/filters/network/ext_authz/ext_authz.cc
Comment (llu94) [reply source/extensions/filters/network/ext_authz/ext_authz.cc:94]: Will discard if null instead of adding tests

File: source/extensions/filters/network/ext_authz/ext_authz.cc
Comment (llu94) [reply source/extensions/filters/network/ext_authz/ext_authz.cc:88]: Will refactor into external function

File: source/common/stream_info/utility.cc
Comment (dio) [inline source/common/stream_info/utility.cc:123]: Nit. Probably choose a different variable name? Since this is not a specific file for ext_authz. Or move it inside the filter code since that is the only one using this?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:223]: Since this "duration" field is set by the filter (not by the ext_authz service), could you help to document it in https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter#dynamic-metadata? Thanks!

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:63]: Seems like this can be a local variable (vs. class member)?

**Review by dio — COMMENTED** [review]
Thanks for working on this. A couple of comments from me:

File: source/common/stream_info/utility.cc
Comment (llu94) [reply source/common/stream_info/utility.cc:123]: Changed the variable name.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (llu94) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Added documentation.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (llu94) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:63]: It can't be a local variable without rewriting what I'm doing since I explicitly reference that variable as the start time in the on complete method.

**Comment by jmarantz** [comment]
@llu94 please do not force-push to github PRs that are in review: https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md has detail about why that's bad and what to do instead.

File: source/common/stream_info/utility.cc
Comment (adisuissa) [inline source/common/stream_info/utility.cc:120]: As @dio said, why is this method in `TimingUtility`?
It should either be a static function called something similar to `timeSince(absl::optional<MonotonicTime> start_time)`, or it can be defined inside the filter code in an empty namespace.

Comment (llu94) [reply source/common/stream_info/utility.cc]: I moved it into an empty namespace.

**Comment by llu94** [comment]
> @llu94 please do not force-push to github PRs that are in review: https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md has detail about why that's bad and what to do instead.

Understood. I am having some issues with the system refusing to pick up my sign offs and had to do this and DCO signoff checks failing. I think I have fixed this though and this won't be a problem in the future.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (adisuissa) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:233]: It seems that the if statement is true iff `start_time_` was set.
In this case, why not have the if statement (`if (start_time.has_value()) {`) here, and then set the number from the `time_since_epoch()` computation?

Comment (llu94) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Changes have been made. Good catch.

**Comment by llu94** [comment]
Code coverage limits not being met. Will fix.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks.
/assign-from @envoyproxy/senior-maintainers

**Comment by llu94** [comment]
@adisuissa I don't have authorization to merge. Can you merge for me?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (zuercher) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:2140]: I'd like to see one of tests (probably one for the http filter and one for the network filter) check that the duration is an expected value. I think you could configure the  connection returned by `filter_callbacks_` in the test class to return an `SimulatedTimeSystem` (it may already) and use that to manipulate the result of the calls to `now` and then check that the duration is the expected value.

**Review by zuercher — CHANGES_REQUESTED** [review]
Thanks! Please see my comments below.

File: docs/root/configuration/http/http_filters/ext_authz_filter.rst
Comment (llu94) [reply docs/root/configuration/http/http_filters/ext_authz_filter.rst:194]: Done.

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (llu94) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc:2140]: I'll need to think a little bit on how to do this.

**Comment by llu94** [comment]
@zuercher I have made some changes to the test in line with your suggestions using a simulated time system.

File: test/extensions/filters/network/ext_authz/ext_authz_test.cc
Comment (zuercher) [inline test/extensions/filters/network/ext_authz/ext_authz_test.cc:47]: Instead, I think we can do this (before calling initializeReadFilterCallbacks):

```
auto time_system = std::make_unique<Event::SimulatedTimeSystem>();
time_system->setMonotonicTime(std::chrono::milliseconds(0));
time_system_ = time_system.get();
filter_callbacks_->connection_.dispatcher_.time_system_ = time_system;
```
(The `time_system_` field will need to be made into a pointer type.)

That makes the time system used in all these tests the simulated time you want and keeps a pointer to the time system available in the tests (just be careful not to use it when its invalid).

Then you shouldn't need to expose `setStartTime` or `getStartTime` in your filters.

File: test/extensions/filters/network/ext_authz/ext_authz_test.cc
Comment (zuercher) [inline test/extensions/filters/network/ext_authz/ext_authz_test.cc:106]: I think we can dispense with creating duration and just put `10` here.

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (zuercher) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:63]: Similar to the network filter test, we can create a time system and assign it to `decoder_filter_callbacks_->dispatcher_.time_system_`.

File: source/extensions/filters/http/ext_authz/ext_authz.h
Comment (zuercher) [inline source/extensions/filters/http/ext_authz/ext_authz.h:268]: See my comments in the tests, but I don't think we want to add these public methods.

**Review by zuercher — APPROVED** [review]
Thanks!

File: test/extensions/filters/network/ext_authz/ext_authz_test.cc
Comment (llu94) [reply test/extensions/filters/network/ext_authz/ext_authz_test.cc]: Did some digging into the MockDispatcher and found a GlobalTimeSystem that we can use instead.

**Comment by llu94** [comment]
No worries. Thanks to you and all of the other reviewers for guiding me through this first PR. I am by no means good with C++ and this code base is pretty daunting at times.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by llu94** [comment]
I can't merge it. Not authorized. Can someone else do it for me?

**Pull request merged**
